### ISSUE TITLE
Help with BIOMOD_EnsembleForecasting - [I can't get my predicted value]

### ISSUE THREAD

**Issue by TY-dispersal** [biomodhub/biomod2#399]
**Context and question**
I used 10 models to build a species distribution model and built ensemble  model based on the model results. I use ensemble models to make predictions of potential distributions in the context of future and past climates. In future predictions, there was a situation where the prediction results could not be printed using the plot() function, and in past predictions I could not get the predicted values.

**Code used**
file_paths <- c(.....)
interest_area <- extent(115, 165, 30, 65)
myExpl_clipped <- terra::crop(terra::rast(file_paths), interest_area)
point <- read.csv(...)
myRespName <- 'PK'
myResp <- as.numeric(point[, 3])
myRespXY <- point[, c('Longitude', 'Latitude')]
path_to_maxent.jar <- "E:/maxent/maxent/maxent.jar"
myBiomodData <- 
  BIOMOD_FormatingData(resp.var = myResp,
                       expl.var = myExpl_clipped,
                       resp.xy = myRespXY,
                       resp.name = myRespName,
                       PA.nb.rep = 2,
                       PA.strategy = 'random',
                       PA.nb.absences = 1000)
plot(myBiomodData)
myBiomodOptions <- BIOMOD_ModelingOptions(MAXENT = list(path_to_maxent.jar = path_to_maxent.jar))
bm.tuning <- BIOMOD_Tuning(
  bm.format = myBiomodData,
  models = c("GLM", "GBM", "GAM", "CTA", "ANN", "FDA", "MARS", "RF"),
  bm.options = myBiomodOptions,
  metric.eval = "ROC",
  ctrl.train.tuneLength = 30,
  ANN.decay.tune = c(0.001, 0.01, 0.05, 0.1),
  ANN.size.tune = c(2, 4, 6, 8),
  ANN.maxit = 500,
  ANN.MaxNWts = 10 * (ncol(myBiomodData@data.env.var) + 1) + 10 + 1,
  MARS.method = "earth",
  GAM.method = "gam",
  GLM.method = "glmStepAIC",
  GLM.type = c("simple", "quadratic", "polynomial", "s_smoother"),
  GLM.interaction = c(0, 1),
  ME.cvmethod = "randomkfold",
  ME.kfolds = 10,
  ME.n.bg = 10000,
  ME.metric = "ROC",
  ME.parallel = FALSE)
myBiomodOptions <- bm.tuning$models.options
myBiomodModelOut <- 
  BIOMOD_Modeling(bm.format = myBiomodData,
                  modeling.id = 'AllModels',
                  models = c("GLM", "GBM", "GAM", "CTA", "ANN", "SRE", "FDA", "MARS", "RF", "MAXENT"),
                  bm.options = myBiomodOptions,
                  CV.strategy = 'random',
                  CV.nb.rep = 10,
                  CV.perc = 0.75,
                  metric.eval = c('TSS','ROC'),
                  var.import = 2,
                  prevalence = 0.5,
                  seed.val = 42)
myBiomodEM <- 
  BIOMOD_EnsembleModeling(bm.mod = myBiomodModelOut,
                          models.chosen = 'all',
                          em.by = 'all',
                          em.algo = c('EMwmean'),
                          metric.select = c('TSS'),
                          metric.select.thresh = c(0.7),
                          metric.eval = c('TSS', 'ROC'),
                          var.import = 3,
                          seed.val = 42)
myBiomodEMProj <- 
  BIOMOD_EnsembleForecasting(bm.em = myBiomodEM,
                             proj.name = 'CurrentEM',
                             new.env = myExpl_clipped,
                             models.chosen = 'all',
                             metric.binary = 'all',
                             metric.filter = 'all')
####I haven't encountered any obvious problems up to this point (except：crs do not match)
climate_file_path4 <- "E:/climate-2.5′/future2.5-CMCC-ESM2_ssp585.tif"
future_env_file4 <- terra::rast(climate_file_path4)
indices <- c(2, 3, 4, 5, 12, 15)
selected_climate_factors4 <- future_env_file4[[indices]]
names(selected_climate_factors4) <- c("bio2", "bio3", "bio4", "bio5", "bio12", "bio15")
selected_climate_factors4 <- crop(selected_climate_factors4, interest_area)
future_env_factors4 <- c(selected_climate_factors4, elev, Aspect, Slope)
myBiomodEMProj4 <- 
  BIOMOD_EnsembleForecasting(bm.em = myBiomodEM,
                             proj.name = 'future4',
                             new.env = future_env_factors4,
                             models.chosen = 'all',
                             metric.binary = 'all',
                             metric.filter = 'all')
spatialPredictions4 <- myBiomodEMProj4@proj.out
packedRaster4 <- spatialPredictions4@val
standardRaster4 <- rast(packedRaster4)
writeRaster(standardRaster4, "CMCC-ESM2_ssp585.tif", overwrite=TRUE)
####Here I can still get the predicted result and export it and have the predicted value. It's just that I can't use plot()
> plot(myBiomodEMProj4)
wrong: [rast] file does not exist: C:/Users/lty/AppData/Local/Temp/RtmpyicLUG/spat_7wSvXHEjr5EdmV7_1224.tif
moreover : Warning message:
C:/Users/lty/AppData/Local/Temp/RtmpyicLUG/spat_7wSvXHEjr5EdmV7_1224.tif: No such file or directory (GDAL error 4) 
##past
PASTMHMIROC_path <- c(...)
PASTMHMIROC <- terra::rast(PASTMHMIROC_path)
PASTMHMIROC <- terra::crop(PASTMHMIROC, interest_area)
terrain_data <- c(elev, Aspect, Slope)#elev, Aspect, Slope are already cropped raster files
crs_terrain <- crs(terrain_data)
climate_data_projected <- terra::project(PASTMHMIROC, crs_terrain)
PASTMHMIROC1 <- terra::resample(climate_data_projected, terrain_data)
PASTMHMIROC2 <- c(PASTMHMIROC1, terrain_data)
myBiomodEMProj7 <- 
  BIOMOD_EnsembleForecasting(bm.em = myBiomodEM,
                             proj.name = 'pastMHmir',
                             new.env = PASTMHMIROC2,
                             models.chosen = 'all',
                             metric.binary = 'all',
                             metric.filter = 'all')
str(myBiomodEMProj7)
spatialPredictions7 <- myBiomodEMProj7@proj.out
packedRaster7 <- spatialPredictions7@val
standardRaster7 <- rast(packedRaster7)
writeRaster(standardRaster7, "MH-MIROC4.tif", overwrite=TRUE)
####Here I still completed the prediction of the model and got the output results. But there is no predictive value in this result.
> standardRaster7
class       : SpatRaster 
dimensions  : 840, 1200, 1  (nrow, ncol, nlyr)
resolution  : 0.04166667, 0.04166667  (x, y)
extent      : 115, 165, 30, 65  (xmin, xmax, ymin, ymax)
coord. ref. : lon/lat WGS 84 (EPSG:4326) 
source(s)   : memory
name        : PK_EMwmeanByTSS_mergedData_mergedRun_mergedAlgo 
min value   :                                             NaN 
max value   :                                             NaN 
I have tried reading environment data many times, including simple methods，such as：
PASTMHMIROC <- terra::rast(PASTMHMIROC_path)
PASTMHMIROC2 <- c(PASTMHMIROC, elev, Aspect, Slope)
But I don't get the predicted value

#################################################
####Below is some of my model result output
> show(myExpl_clipped)
class       : SpatRaster 
dimensions  : 840, 1200, 9  (nrow, ncol, nlyr)
resolution  : 0.04166667, 0.04166667  (x, y)
extent      : 115, 165, 30, 65  (xmin, xmax, ymin, ymax)
coord. ref. : lon/lat WGS 84 (EPSG:4326) 
source(s)   : memory
names       :     bio2,      bio3,     bio4,        bio5, bio12,     bio15, ... 
min values  :  1.00000,  10.27035,    0.000,  0.03999999,   188,  13.52208, ... 
max values  : 18.13567, 100.00000, 2377.624, 33.81999969,  3490, 152.96664, ...
<html>
<body>

> show(myBiomodData) 
 -=-=-=-=-=-=-=-=-=-=-= BIOMOD.formated.data -=-=-=-=-=-=-=-=-=-=-=  dir.name =  .  sp.name =  PK  	 738 presences,  0 true absences and  1999  undefined points in dataset   	 9 explanatory variables        bio2            bio3            bio4             bio5        Min.   : 5.03   Min.   :13.47   Min.   : 617.1   Min.   :10.03    1st Qu.:10.56   1st Qu.:21.11   1st Qu.:1243.7   1st Qu.:22.21    Median :12.00   Median :23.73   Median :1482.7   Median :24.32    Mean   :11.69   Mean   :23.36   Mean   :1477.1   Mean   :23.95    3rd Qu.:13.00   3rd Qu.:25.75   3rd Qu.:1707.7   3rd Qu.:26.16    Max.   :17.89   Max.   :31.46   Max.   :2349.9   Max.   :33.01        bio12            bio15             elev         Min.   : 190.0   Min.   : 14.56   Min.   :  -3.0    1st Qu.: 411.0   1st Qu.: 64.58   1st Qu.: 224.0    Median : 581.0   Median : 82.75   Median : 444.0    Mean   : 644.2   Mean   : 81.30   Mean   : 527.5    3rd Qu.: 732.0   3rd Qu.:100.01   3rd Qu.: 767.0    Max.   :3268.0   Max.   :141.26   Max.   :2325.0        Aspect           Slope         Min.   : -1.00   Min.   :0.0000    1st Qu.: 94.18   1st Qu.:0.3412    Median :175.46   Median :0.8264    Mean   :180.13   Mean   :1.0778    3rd Qu.:270.48   3rd Qu.:1.4807    Max.   :359.73   Max.   :9.5791      2 Pseudo Absences dataset available ( PA1, PA2 ) with   1000 (PA1, PA2) pseudo absences  -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
> show(myBiomodModelOut)
-=-=-=-=-=-=-=-=-=-=-=-= BIOMOD.models.out -=-=-=-=-=-=-=-=-=-=-=-=
Modeling folder : .
Species modeled : PK
Modeling id : AllModels
Considered variables : bio2 bio3 bio4 bio5 bio12 bio15 elev 
Aspect Slope
Computed Models :  PK_PA1_RUN1_GLM PK_PA1_RUN1_GBM 
PK_PA1_RUN1_GAM PK_PA1_RUN1_CTA PK_PA1_RUN1_ANN PK_PA1_RUN1_SRE 
PK_PA1_RUN1_FDA PK_PA1_RUN1_MARS PK_PA1_RUN1_RF PK_PA1_RUN1_MAXENT 
PK_PA1_RUN2_GLM PK_PA1_RUN2_GBM PK_PA1_RUN2_GAM PK_PA1_RUN2_CTA 
PK_PA1_RUN2_ANN PK_PA1_RUN2_SRE PK_PA1_RUN2_FDA PK_PA1_RUN2_MARS 
PK_PA1_RUN2_RF PK_PA1_RUN2_MAXENT PK_PA1_RUN3_GLM PK_PA1_RUN3_GBM 
PK_PA1_RUN3_GAM PK_PA1_RUN3_CTA PK_PA1_RUN3_SRE PK_PA1_RUN3_FDA 
PK_PA1_RUN3_MARS PK_PA1_RUN3_RF PK_PA1_RUN3_MAXENT PK_PA1_RUN4_GLM 
PK_PA1_RUN4_GBM PK_PA1_RUN4_GAM PK_PA1_RUN4_CTA PK_PA1_RUN4_ANN 
PK_PA1_RUN4_SRE PK_PA1_RUN4_FDA PK_PA1_RUN4_MARS PK_PA1_RUN4_RF 
PK_PA1_RUN4_MAXENT PK_PA1_RUN5_GLM PK_PA1_RUN5_GBM PK_PA1_RUN5_GAM 
PK_PA1_RUN5_CTA PK_PA1_RUN5_ANN PK_PA1_RUN5_SRE PK_PA1_RUN5_FDA 
PK_PA1_RUN5_MARS PK_PA1_RUN5_RF PK_PA1_RUN5_MAXENT PK_PA1_RUN6_GLM 
PK_PA1_RUN6_GBM PK_PA1_RUN6_GAM PK_PA1_RUN6_CTA PK_PA1_RUN6_ANN 
PK_PA1_RUN6_SRE PK_PA1_RUN6_FDA PK_PA1_RUN6_MARS PK_PA1_RUN6_RF 
PK_PA1_RUN6_MAXENT PK_PA1_RUN7_GLM PK_PA1_RUN7_GBM PK_PA1_RUN7_GAM 
PK_PA1_RUN7_CTA PK_PA1_RUN7_ANN PK_PA1_RUN7_SRE PK_PA1_RUN7_FDA 
PK_PA1_RUN7_MARS PK_PA1_RUN7_RF PK_PA1_RUN7_MAXENT PK_PA1_RUN8_GLM 
PK_PA1_RUN8_GBM PK_PA1_RUN8_GAM PK_PA1_RUN8_CTA PK_PA1_RUN8_ANN 
PK_PA1_RUN8_SRE PK_PA1_RUN8_FDA PK_PA1_RUN8_MARS PK_PA1_RUN8_RF 
PK_PA1_RUN8_MAXENT PK_PA1_RUN9_GLM PK_PA1_RUN9_GBM PK_PA1_RUN9_GAM 
PK_PA1_RUN9_CTA PK_PA1_RUN9_ANN PK_PA1_RUN9_SRE PK_PA1_RUN9_FDA 
PK_PA1_RUN9_MARS PK_PA1_RUN9_RF PK_PA1_RUN9_MAXENT 
PK_PA1_RUN10_GLM PK_PA1_RUN10_GBM PK_PA1_RUN10_GAM 
PK_PA1_RUN10_CTA PK_PA1_RUN10_ANN PK_PA1_RUN10_SRE 
PK_PA1_RUN10_FDA PK_PA1_RUN10_MARS PK_PA1_RUN10_RF 
PK_PA1_RUN10_MAXENT PK_PA2_RUN1_GLM PK_PA2_RUN1_GBM 
PK_PA2_RUN1_GAM PK_PA2_RUN1_CTA PK_PA2_RUN1_ANN PK_PA2_RUN1_SRE 
PK_PA2_RUN1_FDA PK_PA2_RUN1_MARS PK_PA2_RUN1_RF PK_PA2_RUN1_MAXENT 
PK_PA2_RUN2_GLM PK_PA2_RUN2_GBM PK_PA2_RUN2_GAM PK_PA2_RUN2_CTA 
PK_PA2_RUN2_ANN PK_PA2_RUN2_SRE PK_PA2_RUN2_FDA PK_PA2_RUN2_MARS 
PK_PA2_RUN2_RF PK_PA2_RUN2_MAXENT PK_PA2_RUN3_GLM PK_PA2_RUN3_GBM 
PK_PA2_RUN3_GAM PK_PA2_RUN3_CTA PK_PA2_RUN3_ANN PK_PA2_RUN3_SRE 
PK_PA2_RUN3_FDA PK_PA2_RUN3_MARS PK_PA2_RUN3_RF PK_PA2_RUN3_MAXENT 
PK_PA2_RUN4_GLM PK_PA2_RUN4_GBM PK_PA2_RUN4_GAM PK_PA2_RUN4_CTA 
PK_PA2_RUN4_SRE PK_PA2_RUN4_FDA PK_PA2_RUN4_MARS PK_PA2_RUN4_RF 
PK_PA2_RUN4_MAXENT PK_PA2_RUN5_GLM PK_PA2_RUN5_GBM PK_PA2_RUN5_GAM 
PK_PA2_RUN5_CTA PK_PA2_RUN5_ANN PK_PA2_RUN5_SRE PK_PA2_RUN5_FDA 
PK_PA2_RUN5_MARS PK_PA2_RUN5_RF PK_PA2_RUN5_MAXENT PK_PA2_RUN6_GLM 
PK_PA2_RUN6_GBM PK_PA2_RUN6_GAM PK_PA2_RUN6_CTA PK_PA2_RUN6_ANN 
PK_PA2_RUN6_SRE PK_PA2_RUN6_FDA PK_PA2_RUN6_MARS PK_PA2_RUN6_RF 
PK_PA2_RUN6_MAXENT PK_PA2_RUN7_GLM PK_PA2_RUN7_GBM PK_PA2_RUN7_GAM 
PK_PA2_RUN7_CTA PK_PA2_RUN7_ANN PK_PA2_RUN7_SRE PK_PA2_RUN7_FDA 
PK_PA2_RUN7_MARS PK_PA2_RUN7_RF PK_PA2_RUN7_MAXENT PK_PA2_RUN8_GLM 
PK_PA2_RUN8_GBM PK_PA2_RUN8_GAM PK_PA2_RUN8_CTA PK_PA2_RUN8_ANN 
PK_PA2_RUN8_SRE PK_PA2_RUN8_FDA PK_PA2_RUN8_MARS PK_PA2_RUN8_RF 
PK_PA2_RUN8_MAXENT PK_PA2_RUN9_GLM PK_PA2_RUN9_GBM PK_PA2_RUN9_GAM 
PK_PA2_RUN9_CTA PK_PA2_RUN9_ANN PK_PA2_RUN9_SRE PK_PA2_RUN9_FDA 
PK_PA2_RUN9_MARS PK_PA2_RUN9_RF PK_PA2_RUN9_MAXENT 
PK_PA2_RUN10_GLM PK_PA2_RUN10_GBM PK_PA2_RUN10_GAM 
PK_PA2_RUN10_CTA PK_PA2_RUN10_ANN PK_PA2_RUN10_SRE 
PK_PA2_RUN10_FDA PK_PA2_RUN10_MARS PK_PA2_RUN10_RF 
PK_PA2_RUN10_MAXENT
Failed Models :  PK_PA1_RUN3_ANN PK_PA2_RUN4_ANN
> show(myBiomodEM)
  -=-=-=-=-=-=-=-=-=-= BIOMOD.ensemble.models.out -=-=-=-=-=-=-=-=-=-=  
sp.name : PK  
expl.var.names : bio2 bio3 bio4 bio5 bio12 bio15 elev Aspect  Slope   
models computed: PK_EMwmeanByTSS_mergedData_mergedRun_mergedAlgo  
models failed: none  
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
> show(myBiomodEMProj) 
 -=-=-=-=-=-=-=-=-=-=-= BIOMOD.projection.out -=-=-=-=-=-=-=-=-=-=-=  
Projection directory : ./PK/CurrentEM  
 sp.name : PK  
expl.var.names : bio2 bio3 bio4 bio5 bio12 bio15 elev Aspect  Slope   
modeling.id : AllModels ( ./PK/PK.AllModels.ensemble.models.out ) 
 models.projected :  PK_EMwmeanByTSS_mergedData_mergedRun_mergedAlgo  
available binary projection : TSS, ROC  
available filtered projection : TSS, ROC  -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
> show(myBiomodEMProj4)  
-=-=-=-=-=-=-=-=-=-=-= BIOMOD.projection.out -=-=-=-=-=-=-=-=-=-=-= 
 Projection directory : ./PK/future4  
 sp.name : PK  
expl.var.names : bio2 bio3 bio4 bio5 bio12 bio15 elev Aspect  Slope  
 modeling.id : AllModels ( ./PK/PK.AllModels.ensemble.models.out ) 
 models.projected :  PK_EMwmeanByTSS_mergedData_mergedRun_mergedAlgo 
 available binary projection : TSS, ROC  
available filtered projection : TSS, ROC  -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
> show(future_env_factors4)
class       : SpatRaster 
dimensions  : 840, 1200, 9  (nrow, ncol, nlyr)
resolution  : 0.04166667, 0.04166667  (x, y)
extent      : 115, 165, 30, 65  (xmin, xmax, ymin, ymax)
coord. ref. : lon/lat WGS 84 (EPSG:4326) 
source(s)   : memory
varnames    : future2.5-CMCC-ESM2_ssp585 
              elev 
              Aspect 
              ...
names       : bio2,  bio3,   bio4, bio5, bio12, bio15, ... 
min values  : -0.7, -10.4,   36.3,  7.0,   215,  18.4, ... 
max values  : 18.4,  39.4, 2293.0, 39.4,  3339, 162.9, ... 
> show(PASTMHMIROC2)
class       : SpatRaster 
dimensions  : 840, 1200, 9  (nrow, ncol, nlyr)
resolution  : 0.04166667, 0.04166667  (x, y)
extent      : 115, 165, 30, 65  (xmin, xmax, ymin, ymax)
coord. ref. : lon/lat WGS 84 (EPSG:4326) 
source(s)   : memory
names       : bio2, bio3,  bio4, bio5, bio12, bio15, ... 
min values  :   48,   13,  4743,  -38,   141,    20, ... 
max values  :  167,   32, 24006,  362,  3944,   138, ... 
> show(myBiomodEMProj7)
-=-=-=-=-=-=-=-=-=-=-= BIOMOD.projection.out -=-=-=-=-=-=-=-=-=-=-=
Projection directory : ./PK/pastMHmir
sp.name : PK
expl.var.names : bio2 bio3 bio4 bio5 bio12 bio15 elev Aspect 
Slope
modeling.id : AllModels ( ./PK/PK.AllModels.ensemble.models.out )
models.projected : 
PK_EMwmeanByTSS_mergedData_mergedRun_mergedAlgo
available binary projection : TSS, ROC
available filtered projection : TSS, ROC
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
I really like using the biomod2 package. Although I am a beginner in R, I started using R when building models using the biomod2 package. So I really need your help.

**Comment by HeleneBlt** [comment]
Hello there, 

Thanks for reporting 🙏 
For the first error, for the plot of `myBiomodEMProj4`, it seems there is a problem with the temporary files. Are you able to plot your other rasters as `myExpl_clipped `or `future_env_factors4 `? One of the first things to do is to uninstall / reinstall terra (and maybe biomod2 if necessary) with the command `remove.packages("terra")` and `install.packages("terra")`. 

For the NA in `standardRaster7`, it seems some NA in `PASTMHMIROC2` disturbed the model. Could you have a look at where there is NA? I also noticed some of your variables (bio2, bio4) have a completely different range in PASTMHMIROC2 than for current or future conditions: is it normal? 

Hélène

**Comment by TY-dispersal** [comment]
Thank you very much for your reply.

myBiomodEMProj4 is one of my predictions. I don't know what causes this. Some results of BIOMOD_EnsembleForecasting() can be plotted (), but some cannot. The solution is to run BIOMOD_EnsembleForecasting() again and the results obtained can be plotted (). Moreover, while I cannot plot() after running BIOMOD_EnsembleForecasting(), my environment data (new.env) cannot be plot() viewed. I restarted Rstudio and it couldn't solve the problem. I can only view the environment data by re-importing it.

There are NA's in my environment data. This is because part of the range I captured is the ocean. There are no environmental data values in the ocean.

I checked many times. I discovered what might be causing the lack of predicted values in my predictions. In the bioclimatic data I downloaded (bio), the data on temperature seems to be amplified by a factor of 10. That is to say, it is very different from the value of the environmental data I used when training the model. I divided the past data/10 and got the value in the normal range. I now successfully get the predicted results.
It’s just that at the end of my predictions, a warning appears: crs do not match.

Thank you again
